File: server-dotnet/src/Ecommerce.Domain/Models/Product.cs
namespace Ecommerce.Domain.Models;

public sealed class Product
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public decimal BasePrice { get; set; }
    public string? CountryOfOrigin { get; set; }
    public int StockQuantity { get; set; }
    public int? ManufacturerId { get; set; }
    public string? Category { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public Manufacturer? Manufacturer { get; set; }
    public List<ProductImage> Images { get; set; } = [];
    public List<ProductStatus> Statuses { get; set; } = [];
}

public sealed class ProductImage
{
    public int Id { get; set; }
    public int ProductId { get; set; }
    public string Url { get; set; } = string.Empty;
    public bool IsMain { get; set; }
    public int SortOrder { get; set; }
    public Product Product { get; set; } = null!;
}

public sealed class ProductStatus
{
    public int Id { get; set; }
    public int ProductId { get; set; }
    public string StatusType { get; set; } = string.Empty;
    public Product Product { get; set; } = null!;
}

File: client/src/types/product.ts
export interface ProductImageDtoType {
  id: number;
  url: string;
  is_main: boolean;
}

export interface ProductDtoType {
  id: number;
  name: string;
  description: string | null;
  base_price: number;
  country_of_origin: string | null;
  stock_quantity: number;
  manufacturer_id: number | null;
  images: ProductImageDtoType[];
  statuses: string[];
}

// Frontend-friendly product format
export interface Product {
  id: number;
  name: string;
  description: string | null;
  price: number; // mapped from base_price
  country_of_origin: string | null;
  inStock: boolean; // mapped from stock_quantity > 0
  stock_quantity: number;
  manufacturer_id: number | null;
  images: ProductImageDtoType[];
  mainImage?: string; // first image or main image URL
  statuses: string[];
}

export function mapProductDtoToProduct(dto: ProductDtoType): Product {
  const mainImage = dto.images.find(img => img.is_main)?.url || dto.images[0]?.url || '/placeholder.svg';
  
  // Parse statuses if it's a JSON string, otherwise use as array
  let statuses: string[] = [];
  if (dto.statuses) {
    if (typeof dto.statuses === 'string') {
      try {
        statuses = JSON.parse(dto.statuses);
      } catch {
        statuses = [];
      }
    } else if (Array.isArray(dto.statuses)) {
      statuses = dto.statuses;
    }
  }
  
  return {
    id: dto.id,
    name: dto.name,
    description: dto.description,
    price: dto.base_price,
    country_of_origin: dto.country_of_origin,
    inStock: dto.stock_quantity > 0,
    stock_quantity: dto.stock_quantity,
    manufacturer_id: dto.manufacturer_id,
    images: dto.images,
    mainImage,
    statuses,
  };
}

